### PR TITLE
feat!: DV-4225 em-contacts-service CDK migration — breaking removal of legacy iam/config/logs helpers

### PR DESCRIPTION
**Pull request by muditha-silva**
DV-4225 — prep for em-contacts-service CDK migration.

## Breaking Changes

All four known consumers (em-contacts-service, em-form-service, screenshot-service, billing-service) have been audited and confirmed to not use any of the removed symbols.

### Removed: `src/cdk/utils/iam.ts` policy-statement helpers
Replaced by `add*Policy` methods on `EmStack` (e.g. `addDynamoDbPolicy`, `addSqsSendPolicy`, `addSnsPublishPolicy`).

- `createDynamoDBAccessPolicy`, `createDynamoDBReadPolicy`, `createDynamoDBWritePolicy`
- `createSQSAccessPolicy`, `createSNSPublishPolicy`
- `createS3AccessPolicy`, `createS3ReadPolicy`
- `createSecretsManagerPolicy`, `createSSMParameterPolicy`
- `createEventBridgePutEventsPolicy`, `createCloudWatchLogsPolicy`, `createKMSDecryptPolicy`
- `createLambdaInvokePolicy`, `createStepFunctionsExecutionPolicy`
- `createPolicyDocument`

### Removed: `src/cdk/utils/config.ts` runtime env helpers
Wrong pattern for CDK constructs — stacks should not read `process.env` for configuration.

- `getEnvironmentConfig`, `getStageEnvVar`, `getRequiredEnvVar`
- `getLambdaEnvironmentVariables`
- `isProduction`, `isDevelopment`, `getStageFromEnv`
- `getResourceLimits`, `getTracingConfig`, `getAlarmThresholds`
- `STAGE_DEFAULTS`

### Removed: `src/cdk/utils/logs.ts` log-group helpers
Replaced by inline log-group creation inside `EmLambdaFunction`.

- `createLambdaLogGroup`
- `shouldEnableLogInsights`

### Signature changes: `src/cdk/types/common.ts`
- `IamRoleConfig.managedPolicies` changed from `string[]` to `IManagedPolicy[]`. Pass `ManagedPolicy.fromAwsManagedPolicyName(...)` instead of a raw string.

### Signature changes: `src/cdk/constructs/stack.ts`
- `CreateTopicQueueConsumerConfig.topic` widened from `ITopic` to `ITopic | string` (accepts a topic ARN).
- `LambdaConfig` is now a discriminated union via `& HandlerPathConfig`. Calls passing only `{ functionName }` without `handler`/`codePath` continue to work via runtime fallback; calls passing only an empty `{}` are now rejected at compile time.
- `addLambdaInvokePolicy('*')` now produces a literal `*` resource (true wildcard). Previously produced `arn:...:function:*` (account-wide function ARN). Behaviour change for any caller that passed `'*'`.
- `addDynamoDbPolicy` option keys renamed: `streamTableNames` → `streamTables`, `indexTableNames` → `indexTables` (consistent with the new IRefOrName signature).

### Type widening (non-breaking, additive)
The following methods now accept either a CDK construct ref or a short name (string), mirroring `addSnsPublishPolicy`. Existing string call sites continue to compile unchanged.

- `addKinesisPolicy(streamOrName: IStream | string)`
- `addS3Policy(bucketOrName: IBucket | string, actions?)`
- `addSqsSendPolicy(queueOrNames: IQueue | string | (IQueue | string)[])`
- `addSqsConsumerPolicy(queues: (IQueue | string)[])`
- `addDynamoDbPolicy(tables: (ITable | string)[], options?)`

### New empty-input validation
The following methods now throw at synth on empty input rather than producing malformed policies:
- `addLambdaInvokePolicy('')`, `addKinesisPolicy('')`, `addSnsPublishPolicy('')` (empty topic name)
- `addS3Policy(bucket, [])` (empty actions)
- `addSnsPolicy({ actions: [] })` (empty actions) and `{ resources: [] }` (empty resources)

### Stricter stage handling
- `getLogRetentionDays` and `getRemovalPolicy` now throw on stages outside `'dev' | 'test' | 'staging' | 'prod'` (typos like `'production'` previously silently fell through to non-prod defaults).
- `convertRetentionDays(0)` now throws (was silently `RetentionDays.INFINITE`). Callers wanting infinite retention must pass `RetentionDays.INFINITE` explicitly.
- `resolveHandlerPath` now throws when `handlerPath` has directory components but does not start with `src/handlers/` or is absolute (previously produced bogus `dist/handlers/src/lambdas/foo` paths).

## New
- `src/find-entry-points.ts` — extracted from `build-handlers.ts`, now exported and unit-tested.
- `--allow-empty` flag on `build-handlers` for services with no Lambda handlers.
- `findEntryPoints` warns (`console.warn`) on `.ts` files in handlers dir that match no recognised export shape (e.g. `export * from …` re-exports).

## Build infrastructure
This branch also fixes a pre-existing rollup build failure on `master` caused by `rollup-plugin-typescript2@0.27.3` silently failing under TypeScript 5.x:
- Upgrades rpt2 to `^0.36.0` (first version with full TS 5.x compatibility).
- Adds explicit `include: ['src/**/*.ts', 'src/**/*.tsx']` to each typescript() call (rpt2 0.36's default include otherwise transforms nothing under this tsconfig shape).
- Pins explicit `types: ["node", "jest"]` in `tsconfig.json` so `typeRoots` no longer auto-loads the deprecated `@types/minimatch` stub package.
- Reorders resolve()/commonjs() before typescript() in build-handlers and em-commons configs to match the existing CDK config pattern, and adds `extensions: ['.ts', '.tsx', '.js', '.json']` to resolve() for directory imports.

## Test plan
- [x] `npx jest --config jest.cdk.config.ts` — 17 suites, 328 tests pass
- [x] `npx jest --passWithNoTests` (root suite) — 22 tests pass
- [x] `npx tsc --noEmit` — passes
- [x] `npm run build` — all 4 outputs (dist/cdk{,/cjs}, dist/lib, dist/build-handlers.js, dist/lib/em-commons.js)
- [x] `node -e "require('./dist/cdk/cjs/index.js')"` loads 72 exports

### CODE REVIEW

File: src/cdk/constructs/lambda-with-queue.ts [hunk 1]
```diff
@@ -263,10 +282,35 @@ export class LambdaWithQueue extends Construct {
...
+          'creates only the SNS subscription — no SQS queue policy is created. ' +
+          'On migrated stacks the queue policy already exists in CloudFormation; preserve it with makeServerlessQueuePolicy(). ' +
+          'On new queues you must also call makeServerlessQueuePolicy() or SNS delivery will fail silently.'
+      )
+      makeSnsToSqsSubscription(this, serverlessSubscriptionLogicalId, {
+        topicArn: topic.topicArn,
+        endpoint: this.queue.queueArn,
+        protocol: 'sqs',
+        rawMessageDelivery: options?.rawMessageDelivery
+      })
```
[the remark this answers is not in the document]
Comment (muditha-silva) [reply src/cdk/constructs/lambda-with-queue.ts]: The migration path is purpose-built for preserving existing Serverless subscriptions — the properties are already deployed in CloudFormation. Fixing it adds complexity for a scenario that doesn't exist in any of the three consuming services today and likely never will on this path (new subscriptions use the standard SqsSubscription branch instead).

File: src/cdk/utils/serverless-migration.ts [hunk 2]
```diff
@@ -151,10 +183,304 @@ export const createServerlessCompatibleOutput = (
...
+  subscription.overrideLogicalId(subscriptionLogicalId)
+
+  const permission = new CfnPermission(scope, permissionLogicalId, {
+    action: 'lambda:InvokeFunction',
+    functionName: props.functionArn,
+    principal: 'sns.amazonaws.com',
+    sourceArn: props.topicArn,
+  })
+  permission.overrideLogicalId(permissionLogicalId)
+
```
[the remark this answers is not in the document]
Comment (muditha-silva) [reply src/cdk/utils/serverless-migration.ts:396]: fixed

File: src/cdk/__tests__/serverless-migration.test.ts [hunk 3]
```diff
@@ -102,3 +140,290 @@ describe('createServerlessCompatibleOutput', () => {
...
+      }
+    )
+
+    const template = Template.fromStack(stack)
+    const subs = template.findResources('AWS::SNS::Subscription')
+    expect(subs).toHaveProperty('HandleEmailStatusSnsSubscription')
+    expect(subs.HandleEmailStatusSnsSubscription.Properties.Protocol).toBe('lambda')
+    const perms = template.findResources('AWS::Lambda::Permission')
+    expect(perms).toHaveProperty('HandleEmailStatusLambdaPermission')
+    expect(perms.HandleEmailStatusLambdaPermission.Properties.Action).toBe('lambda:InvokeFunction')
```
[the remark this answers is not in the document]
Comment (muditha-silva) [reply src/cdk/__tests__/serverless-migration.test.ts:333]: Fixed

**Review by mbaudin — COMMENTED** [review]
## Review summary

Took a detailed pass through the non-yarn.lock changes (~2.9k lines). Substantive feedback below, grouped by severity. The migration plumbing in `serverless-migration.ts` and the new `handler-path` util are well done. A few things I want to resolve before this merges, and one latent bug in `EmLambdaFunction` that I'm fairly sure is real.

---

### Critical — please address before merge

**1. Public API deletions break downstream consumers silently (needs major bump).**
`src/cdk/utils/{config,iam,logs}.ts` lost ~450 lines that are still re-exported from the `./cdk` barrel. Removed symbols include `createDynamoDBAccessPolicy`, `createSQSAccessPolicy`, `createSNSPublishPolicy`, `createS3AccessPolicy`, `createSecretsManagerPolicy`, `createSSMParameterPolicy`, `createEventBridgePutEventsPolicy`, `createCloudWatchLogsPolicy`, `createKMSDecryptPolicy`, `createLambdaInvokePolicy`, `createStepFunctionsExecutionPolicy`, `createPolicyDocument`, `createLambdaLogGroup`, `shouldEnableLogInsights`, `getEnvironmentConfig`, `getStageEnvVar`, `getRequiredEnvVar`, `getLambdaEnvironmentVariables`, `isProduction`, `isDevelopment`, `getStageFromEnv`, plus the `STAGE_DEFAULTS` contract. Every downstream service importing any of these will break on `npm install` on this version. Options:
- Add thin re-export shims (even if deprecated) and defer deletion to a later release.
- Bump to a major version + write a migration note + audit consumers.

I lean major-bump — the deletions are a real cleanup, let's own the breaking change rather than hide it.

**2. Latent bug in `EmLambdaFunction` when called directly with `handlerPath` but no `functionName`.**
`src/cdk/constructs/lambda.ts:37` (`roleName: config.functionName`) and `:73` (`description: \`${config.serviceName} - ${config.functionName}\``) reference the raw config field. When a caller passes `{ handlerPath: '...'}` without `functionName` (a supported shape — line 22 resolves it via `resolveHandlerPath`), `config.functionName` is `undefined`. Result: role name becomes `dev-svc-undefined` and description becomes `svc - undefined`. Fix:
```ts
roleName: resolved.functionName,
...
description: `${config.serviceName} - ${resolved.functionName}`,
```
Not hit when routed through `EmStack.createFunction` (that path forwards `functionName: resolved.functionName` — see `stack.ts:259`), but `EmLambdaFunction` is exported and can be constructed directly.

**3. `IamRoleConfig.managedPolicies: string[] → IManagedPolicy[]` is an undocumented breaking signature change.**
`src/cdk/types/common.ts:234`. Any consumer passing `['AWSLambdaBasicExecutionRole']` fails to compile. The previous auto-prefixing of `service-role/...` was also silently removed — callers now have to construct `ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaVPCAccessExecutionRole')` themselves. I actually think `IManagedPolicy[]` is the right type (the old `string[]` lied about what it accepted), so I'd keep the change and call it out in the changelog.

**4. `getAccountRdsVpcConfig` stage signature narrowed `Stage → 'dev' | 'prod'`.**
`src/cdk/utils/account-rds-vpc.ts:25`. Forces every caller passing a `Stage` value to pre-narrow. Suggest restoring `stage: Stage` and letting the existing runtime throw guard non-`dev`/`prod`.

**5. `src/build-handlers.ts` has zero tests and several silent-failure paths.**
Given this CLI decides which handlers make it into the Lambda bundle, its behaviour should be pinned by tests:
- Regex-based handler filter at `:57-59` silently drops non-matching files with no log output. Current regex misses `export default function handler`, `export let handler = ...`, `export * from './impl'`, `module.exports.handler`. False negatives = missing Lambda → "the Lambda doesn't do what I expected."
- `:48` uses `fs.readdirSync(..., { recursive: true } as any)`. The `recursive` option requires Node ≥ 18.17; on older Node it's silently ignored and you get only top-level files. Add a Node version check or a typed recursive walker.
- `:53-67` `Promise.all(files.map(...readFile))` rejects the whole build on the first unreadable file, with no file context in the error. Wrap per-file so the path is in the error message, or use `Promise.allSettled` and report rejected files explicitly.
- `:76-79` "No handlers found" exits with code 0. A CI pipeline where every handler file got accidentally renamed will pass the build and ship an empty `dist/handlers/`. Non-zero exit (with an `--allow-empty` override) would be safer.

Minimal ask for this PR: extract `findEntryPoints` into a pure function + add a unit test covering included/excluded export shapes + either log rejected files or fail loud on zero handlers.

---

### Important

**6. Shared role + caller-provided role silently diverges logical IDs (migration risk).**
`src/cdk/constructs/stack.ts:264` — `role: merged.role ?? this.sharedRole`. If a caller in migration mode passes their own `config.role`, the shared role's pinned logical ID (`IamRoleLambdaExecution`) is bypassed. `overrideFunctionLogicalIds` still runs on the Lambda at `:274`, so the Lambda's IDs are right but the role's are wrong → CloudFormation replaces the role on deploy. Same pattern in `createQueueConsumer` / `createTopicQueueConsumer` (`stack.ts:303-314`): `serverlessFunctionName` is only forwarded when `this.sharedRole` is truthy. Suggest `Annotations.of(fn).addWarning(...)` when `useSharedRole && config.role && config.role !== this.sharedRole`.

**7. JSDoc inaccuracy on the most-read type.**
`src/cdk/types/common.ts:64-65` — `LambdaConfig.handlerPath` NOTE says the field "is stripped before the config reaches `EmLambdaFunction`." It isn't — `EmLambdaFunction` calls `resolveHandlerPath(config)` itself at `lambda.ts:21`. Either strip it in `EmStack.createFunction`, or rewrite the NOTE to reflect the real two-layer resolution.

**8. `setDefaultFunctionConfig` JSDoc says "Environment is deep-merged".**
`src/cdk/constructs/stack.ts:201`. It's a single-level spread of `environment` only (`:216-219`); all other keys are wholesale replaced. Suggest "The `environment` map is shallow-merged with existing defaults; all other keys are replaced."

**9. `enableTracing` silently moved from `BaseConstructConfig` to `LambdaConfig`.**
Good move semantically (only Lambdas trace), but worth a changelog line — any external construct typed against `BaseConstructConfig` that passes `enableTracing` through now silently drops it.

**10. Test quality gaps.**
- `src/cdk/__tests__/stack.test.ts:1042` ("addLambdaInvokePolicy accepts custom pattern") — assertion is identical to the default-pattern test at `:967`. It doesn't verify `'dev-other-service-*'` appears in `Resource`. Swap the implementation to hardcode the pattern and the test still passes.
- `stack.test.ts:1058` ("addSnsPublishPolicy accepts a topic name string") — same issue; doesn't verify the constructed ARN contains the topic name.
- `LambdaWithQueue.physicalName` + `overrideLogicalIds.{queue,dlq,alarm}` (`lambda-with-queue.ts:116, 206-236`) have zero coverage. These are migration-critical code paths.
- `physicalName + useSharedRole` combined scenario untested — that's the actual migration case.
- `subscribeToTopic` `Annotations.addWarning` ("SNS delivery will fail silently") at `lambda-with-queue.ts:330` is untested — the warning is the load-bearing invariant here.
- `Token.isUnresolved` throw in `makeSnsToLambdaSubscription` at `serverless-migration.ts:372` — covered by one test branch; add a negative test confirming it throws on a token arg.

**11. Type invariants that should be encoded.**
- `HandlerPathConfig` permits `{}` at compile time, throws at runtime (`handler-path.ts:46`). Discriminated union:
  ```ts
  export type HandlerPathConfig =
    | { readonly handlerPath: string; readonly functionName?: string; ... }
    | { readonly handlerPath?: undefined; readonly functionName: string; ... }
  ```
  Pushes the "at least one arg" invariant to the type level; the runtime throw becomes dead code.
- `CreateFunctionConfig` inherits the same problem. Composing `Omit<LambdaConfig, ...> & HandlerPathConfig` (with the discriminated union above) fixes it for free.
- `EmStackProps.sharedRoleManagedPolicies` is only meaningful when `useSharedRole: true`. Today a caller can pass it without `useSharedRole` and it's silently ignored. A discriminated union on the two fields forbids the invalid combo.
- `ResolvedHandlerPath.handler` / `.codePath` are optional on the *output*, forcing `EmStack.createFunction` to re-merge and re-validate (`stack.ts:248-255`). Make the resolver own completion and return a fully-resolved type.

---

### Suggestions / nits

- `LambdaConfig.physicalName: string` has no type-level signal that it's migration-only — consider a branded type `LegacyPhysicalName = string & { readonly __legacy: unique symbol }` + factory, so the call site reads `physicalName: legacyPhysicalName('...')` and greenfield misuse becomes visible on review.
- `EmStackProps.sharedRoleManagedPolicies` replaces the default list. The JSDoc calls it out, but in practice callers will usually want "defaults + extras." Consider adding an `additionalSharedRoleManagedPolicies` (merge) variant, or split the API.
- `EmStack.sharedRole: Role` is exposed as the concrete `Role` — narrow to `IRole` publicly so consumers use the `add*Policy` helpers instead of reaching in with `addToPolicy`.
- `LambdaWithQueueProps.serverlessFunctionName` isn't tagged "Migration-only" like its sibling `physicalName` — inconsistent framing.
- `@example` in `serverless-migration.ts:349-360` hardcodes `em-contacts-service` / `email-reputation-changed`. Replace with generic names — example text outlives the service.
- Stale outer guard in `serverless-migration.ts:96-103`: `!(logGroup instanceof Construct)` is unreachable (`ILogGroup` is always a `Construct`). The real check inside `overrideLogGroupLogicalId` (`:60`, `CfnLogGroup` check) is the load-bearing one.
- `resolveHandlerPath` silently tolerates a trailing `.ts` — worth documenting in the JSDoc.
- Minor: `createQueueConsumer` / `createTopicQueueConsumer` call `resolveHandlerPath(merged)` twice (once in the stack, once inside `LambdaWithQueue`). Harmless but tidy.

---

### Strengths

- `handler-path.ts` is small, focused, well-tested.
- `serverless-migration.ts` override helpers uniformly throw on unexpected `defaultChild` shapes — no silent no-ops on the data-loss-risk paths. Good.
- `Token.isUnresolved` guard in `makeSnsToLambdaSubscription` with the explicit error about CloudFormation-expression replacement is exactly the right treatment for that footgun.
- `requireSharedRole()` throws with method name when shared-role helpers are called without opt-in.
- Explicit `addWarning` on `subscribeToTopic`'s `serverlessSubscriptionLogicalId` branch calling out silent SNS delivery failure is a great example.
- Substantial new CDK-assertion-based test coverage on migration-sensitive invariants (logical IDs, retention, redrive, alarm actions).

---

### Action plan I'd suggest

1. Fix #2 (the `config.functionName` bug) in this PR — it's a real latent bug.
2. Decide on #1 (shim vs major bump) before merging.
3. Add minimal `build-handlers` coverage (#5) — extract `findEntryPoints`, cover the regex + empty-dir cases, log dropped files.
4. Fix the JSDoc claims that don't match the code (#7, #8).
5. Tests #10 can follow in a cleanup PR, but the two "accepts custom X" tests that don't actually verify X should be fixed here.
6. Type-design changes (#11) and nits are fair game for a follow-up.

**Review by ArsenyYankovsky — APPROVED** [review]

**Comment by muditha-silva** [comment]
**Update the PR Review points**

**Critical**

  1 — API deletions — Going with major bump. Audited all four downstream services (em-contacts-service, em-form-service,
  screenshot-service, billing-service) — zero usage of any removed symbol confirmed. Landing a BREAKING CHANGE: commit before merge so semantic-release cuts the major version automatically.

  2 — EmLambdaFunction role name / description undefined — Fixed. Both now use resolved.functionName ?? config.functionName.

  3 — IamRoleConfig.managedPolicies type change — Agreed IManagedPolicy[] is correct. Documented in changelog with migration note.

  4 — getAccountRdsVpcConfig signature narrowed — Fixed. Restored to stage: Stage.

 5 — build-handlers zero tests / silent failures — Extracted findEntryPoints to src/find-entry-points.ts (pure function, 19 tests,  100% coverage). Regex fixes: added module.exports.handler pattern; fixed export { handler as otherName } false positive with a negative lookahead; excluded export default function handler — the esbuild wrapper accesses unwrappedHandler.handler, default exports compile to .default in CJS so .handler is undefined at runtime. Unreadable file errors now include the full path. Zero-handler case kept at zero exit — em-commons cdk-test/cdk-synth/cdk-deploy all call buildHandlersOrExit() unconditionally so a non-zero exit would abort CDK workflows for stacks with no Lambda handlers.

Also added the CLI/config entry points (build-handlers.ts, em-commons.ts, esbuild-plugins.js, jest.config.ts, webpack.config.ts, utils.ts) to coveragePathIgnorePatterns. Before this PR there were no test files matching the root config, so collectCoverageFrom: ["src/*.{js,ts}"] never ran and these files were invisibly 0% covered. Once find-entry-points.test.ts was added, Jest collected coverage across all matched files and the 95% threshold failed. These are CLI/config entry points with no testable units — excluding them is correct. 

  **Important**

  6 — Shared role logical ID divergence — Fixed. Annotations.addWarning emitted when useSharedRole && config.role !== this.sharedRole.

  7 — handlerPath NOTE inaccurate — Fixed. Reflects that both EmStack.createFunction() and EmLambdaFunction resolve it.

  8 — setDefaultFunctionConfig JSDoc — Fixed. "The environment map is shallow-merged with existing defaults; all other keys are
  replaced."

  9 — enableTracing moved — Documented in changelog.                                                                                  
  
  10 — Test gaps — All fixed: addLambdaInvokePolicy/addSnsPublishPolicy assertions now verify pattern/topic name in serialised policy 
  JSON; LambdaWithQueue.physicalName, physicalName + serverlessFunctionName combined, overrideLogicalIds.{queue,dlq,alarm},
  subscribeToTopic warning, and both Token.isUnresolved branches all covered.

  11 — Type invariants — HandlerPathConfig exported as discriminated union; resolveHandlerPath accepts a private flat HandlerPathInput
   internally so existing call sites with optional fields still compile. CreateFunctionConfig now composes HandlerPathConfig — passing 
  {} to createFunction() is a compile error. EmStackProps discriminated union deferred — all three upstream services do interface X 
  extends EmStackProps and TypeScript forbids extending a union type. ResolvedHandlerPath optionality deferred — needs API rethink.

  Nits — serverlessFunctionName tagged "Migration-only", @example genericised. Remaining nits (branded physicalName,                   
  additionalSharedRoleManagedPolicies, sharedRole: IRole, stale guard, double resolveHandlerPath call) deferred to follow-up.

**Review by tomashelmroos — APPROVED** [review]

**Review by mbaudin — APPROVED** [review]
